app/models/course/forum.rb

```diff
@@ -4,8 +4,46 @@ class Course::Forum < ActiveRecord::Base
 
   belongs_to :course, inverse_of: :forums
   has_many :topics, dependent: :destroy, inverse_of: :forum
+  has_many :posts, class_name: Course::Discussion::Post, through: :topics
   has_many :subscriptions, dependent: :destroy, inverse_of: :forum
 
+  # @!method self.with_topic_count
+  #   Augments all returned records with the number of topics in that forum.
+  scope :with_topic_count, (lambda do
+    joins { topics.outer }.
+      select { 'course_forums.*' }.
+      select { count(topics.id).as(topic_count) }.
+      group { course_forums.id }
+  end)
+
+  # @!method self.with_topic_post_count
+  #   Augments all returned records with the number of topic posts in that forum.
+  scope :with_topic_post_count, (lambda do
+    joins { posts.outer }.
+      select { 'course_forums.*' }.
+      select { count(posts.id).as(topic_post_count) }.
+      group { course_forums.id }
+  end)
+
+  # @!method self.with_topic_view_count
+  #   Augments all returned records with the number of topic views in that forum.
+  scope :with_topic_view_count, (lambda do
+    joins { topics.outer }.
+      joins('LEFT OUTER JOIN course_forum_topic_views ON course_forum_topic_views.
+      topic_id = course_forum_topics.id').
+      select { 'course_forums.*' }.
+      select { count(course_forum_topic_views.id).as(topic_view_count) }.
+      group { course_forums.id }
+  end)
+
+  # Return if a user has subscribed this forum
+  #
+  # @param [User] user The user to check
+  # @return [Boolean] True if the has subscribed this forum
+  def subscribed_by?(user)
+    subscriptions.map(&:user_id).include?(user.try(:id))
```

Comment:
We have eager loaded all the `subscriptions`. But things like `subscriptions.where(user: user).any?` generate new db queries instead of using the eager loaded results. 

So that if we call this function in every _forum partial, there's a N+1 problem.